File: order-producer/src/main/java/com/sahakav/coffee_shop/service/OrderCommandServiceImpl.java
package com.sahakav.coffee_shop.service;

import com.sahakav.coffee_shop.event.AbstractCoffeeOrderEvent;
import com.sahakav.coffee_shop.event.CreateCoffeeOrderEvent;
import com.sahakav.coffee_shop.event.handler.CoffeeEventHandler;
import com.sahakav.coffee_shop.model.CoffeeOrder;
import com.sahakav.coffee_shop.model.CoffeeType;
import com.sahakav.coffee_shop.repository.CoffeeOrderRepository;
import io.vavr.control.Try;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Service;

import java.util.UUID;

@Service
public class OrderCommandServiceImpl implements OrderCommandService {

    private static final Logger LOGGER = LoggerFactory.getLogger(OrderCommandService.class);

    private final CoffeeEventHandler coffeeEventHandler;

    private final CoffeeOrderRepository coffeeOrderRepository;

    public OrderCommandServiceImpl(CoffeeEventHandler coffeeEventHandler, CoffeeOrderRepository coffeeOrderRepository) {
        this.coffeeEventHandler = coffeeEventHandler;
        this.coffeeOrderRepository = coffeeOrderRepository;
    }

    @Override
    public Try<UUID> placeOrder(CoffeeType coffeeType) {
        return Try.success(coffeeType)
                .map(CoffeeOrder::new)
                .map(coffeeOrderRepository::save)
                .peek(coffeeOrder -> LOGGER.info("Creating order " + coffeeOrder))
                .map(CreateCoffeeOrderEvent::new)
                .andThen(coffeeEventHandler::handle)
                .map(AbstractCoffeeOrderEvent::getOrderId);

    }
}

File: order-producer/src/main/scala/com/sahakav/coffee_shop/service/OrderCommandServiceImpl.scala
package com.sahakav.coffee_shop.service

import java.util.{Optional, UUID}

import com.sahakav.coffee_shop.event.CreateCoffeeOrderEvent
import com.sahakav.coffee_shop.event.handler.CoffeeEventHandler
import com.sahakav.coffee_shop.model.{CoffeeOrder, CoffeeOrderState, CoffeeType}
import com.sahakav.coffee_shop.repository.CoffeeOrderRepository
import org.slf4j.LoggerFactory
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.stereotype.Service

import scala.jdk.OptionConverters.RichOptional
import scala.util.{Success, Try}
import scala.jdk.javaapi.OptionConverters.toScala


@Service
class OrderCommandServiceImpl(@Autowired val coffeeEventHandler: CoffeeEventHandler, @Autowired val coffeeOrderRepository: CoffeeOrderRepository) extends OrderCommandService {

  private val LOGGER = LoggerFactory.getLogger(classOf[OrderCommandService])


  override def placeOrder(coffeeType: CoffeeType): Try[UUID] = {
    LOGGER.info("Creating order for type '{}'",coffeeType)
    Success(coffeeType)
      .map(c =>  CoffeeOrder(c))
      .map(order => coffeeOrderRepository.save(order))
      .map(order => CreateCoffeeOrderEvent(order))
      .map(event => coffeeEventHandler.handle(event))

}

  override def getOrderState(orderId: UUID): CoffeeOrderState =

     coffeeOrderRepository.findById(orderId).toScala.map(o => o.orderState).get




}
